clustermesh: allow remote configuration to be changed quickly

oblazek opened the pull request:
In case there was a typo in remote cluster configuration which caused etcd connection to fail it was not possible to change the remote etcd configuration without waiting for initialConnectionTimeout.

See commit messages for detailed desc.

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

```release-note
Fix a bug where users are unable to change a wrong remote etcd configuration
```

giorio94 commented on pkg/clustermesh/remote_cluster.go:
```diff
@@ -285,7 +299,7 @@ func (rc *remoteCluster) onInsert(allocator RemoteIdentityWatcher) {
 
 	rc.remoteConnectionControllerName = fmt.Sprintf("remote-etcd-%s", rc.name)
-	rc.restartRemoteConnection(allocator)
+	go rc.restartRemoteConnection(allocator)
```
I don't think this needs to be changed to run asynchronously, since `restartRemoteConnection` is not blocking. It only configures the controller, which itself runs the `DoFunc` asynchronously:

https://github.com/cilium/cilium/blob/60eb3e62d8a3d69f3d371bca38373365b0e62810/pkg/controller/manager.go#L111

giorio94 commented on pkg/clustermesh/remote_cluster.go:
```diff
@@ -158,17 +168,21 @@ func (rc *remoteCluster) releaseOldConnection() {
 }
 
 func (rc *remoteCluster) restartRemoteConnection(allocator RemoteIdentityWatcher) {
+	rc.releaseOldConnection()
```
If this is moved outside of the `DoFunc`, then it would not be re-executed if the `DoFunc` is executed multiple times, which would happen in case an error is returned.

giorio94 replied:
Please, ignore this comment, since addressed as part of the conversation below.

giorio94 commented on pkg/kvstore/etcd.go:
```diff
@@ -743,6 +743,9 @@ func connectEtcdClient(ctx context.Context, config *client.Config, cfgPath strin
 		select {
+		case <-ctx.Done():
+			handleSessionError(ctx.Err())
+			return
```
Rather than handling the context termination here, I would suggest to set it as part of the configuration passed at client creation time \[1\], so that it is automatically used for the different backend operations. Then, an error should be automatically propagated and handled by the existing logic when the context gets closed.

https://github.com/cilium/cilium/blob/60eb3e62d8a3d69f3d371bca38373365b0e62810/pkg/kvstore/etcd.go#L683

giorio94 commented on pkg/clustermesh/remote_cluster.go:
```diff
@@ -113,6 +117,12 @@ func (rc *remoteCluster) getLogger() *logrus.Entry {
+	// cancel the old context
+	if rc.cancelFunc != nil {
+		rc.cancelFunc()
+	}
```
Could you please elaborate on why canceling the context is required? I would expect that the backend is completely terminated with the `Close` operation below, regardless of the connection status (if not, then I would say that is a bug). Terminating early here would also disrupt the reverse termination ordering of the different components.

oblazek replied:
Good point, but imagine the etcd clients doesn't establish the connection properly.. I.e. the errorChan never gets closed https://github.com/cilium/cilium/blob/master/pkg/kvstore/etcd.go#L749 and the whole thing times out on this line https://github.com/cilium/cilium/blob/master/pkg/kvstore/etcd.go#L754 
So admin has to wait 15min for the connection to change which is not very cool.

oblazek replied:
So that also means that the `DoFunc` never returns during these 15mins (the default timeout) and during that time there is no way to cleanup the connection and if `rc.releaseOldConnection()` is inside the `DoFunc` that cannot be accessed as well

giorio94 replied:
You are right, I overlooked that.

I had again a fresh look at the code, and here are my thoughts:
* I agree with you about the need for the context to be canceled when the controller gets updated, to ensure that the `DoFunc` terminates as soon as possible.
* Still, I wonder if it could make sense to add the context canceling logic to the controller implementation, to be enabled by a new `ControllerParam` (e.g.,` CancelDoFuncOnUpdate`) to preserve the existing behavior in the other cases. I assume it could be a rather general behavior to desire that the `DoFunc` terminates as soon as possible if the controller gets updated. IMHO, it could be handled as part of the`updateParamsLocked` function: https://github.com/cilium/cilium/blob/f9d290c20ee9349da28bf8164ab8e7eb178079b0/pkg/controller/controller.go#L320-L328
* `rc.releaseOldConnection()` should be moved back as the first instruction of the `DoFunc`. Otherwise, you might have a race condition if `restartRemoteConnection` gets called when the `DoFunc` is e.g., somewhere here https://github.com/cilium/cilium/blob/f9d290c20ee9349da28bf8164ab8e7eb178079b0/pkg/clustermesh/remote_cluster.go#L247-L248 (but also a bit earlier, depending on the effect of closing the context on the returned errors). In that case, the struct fields would still be nil and thus not cleaned by `releaseOldConnection()`, leading to a leak. It could also cause other problems since they would be almost immediately set to a non-nil value when the `DoFunc` terminates, while they would be expected to be nil since another run of the DoFunc is then being performed. Instead, if it is part of the `DoFunc`, there is the guarantee that the previous run terminated when a new one is performed and triggers the cleanup.


Does that make sense to you?

oblazek replied:
yes totally! sounds great.. should I try to do that? or do you want to do it yourself? :)

giorio94 replied:
Feel free to go ahead if you'd like

oblazek replied:
done

giorio94 reviewed (changes requested):
Thanks @oblazek for your PR. 

I have a few comments inline. I also wonder how you tested this, since there is apparently an issue with the fs watcher, which causes updates to already existing configuration files not to be correctly detected. I'm working to fix that.

oblazek committed:
etcdclient: use parent ctx and fix logic around it

Previously when etcd client context was not set and client
created with client.New(*config) it would use context.TODO,
without any way to cancel the ctx.

In some cases (e.g. wrong endpoints configuration) it would
 be necessary to cancel the ctx, otherwise the client might
be stucked in a retry loop and in case of remoteConnectionController
there wouldn't be any way stop that.

This commit passes a parent context to the client itself.

This commit also fixes the closing of stopWatch channel and rest
of the resources when ctx is done/cancelled.

If ctx was closed and select statement choosed handling of that
case (<-ctx.Done()) the code would quit the for loop and handling
of <-w.stopWatch would never be possible which means calling
w.Stop() from outside (e.g. releaseOldConnection()) would hang
forever as the wg counter wouldn't be decremented.
Now when the context gets closed, it correctly closes all channels
and decrements the wg counter, just like when handled by
releaseOldConnection func which releases everything.

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

oblazek committed:
controller: add new ControllerParam CancelDoFuncOnUpdate

This commit adds a new ControllerParam CancelDoFuncOnUpdate which
allows cancelling the controller context passed to DoFunc.

This way we can terminate the controller (if stuck in e.g. etcd
client retry loop) and still preserve the existing logic.

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

giorio94 commented on pkg/kvstore/etcd.go:
```diff
@@ -1034,15 +1039,10 @@ reList:
-			case <-e.client.Ctx().Done():
-				return
-			case <-ctx.Done():
-				return
```
Do you think that it could be possible to preserve the handling of these two cases here, falling though to the stopWatch case? That would also require calling `w.Stop` after `w.stopWait.Done()` to ensure that the `stopWatch` channel gets always closed (there is no issue if it had already been stopped, since it is protected by `sync.Once`). Otherwise, the burden of calling `w.Stop()` is moved to each caller of `ListAndWatch`, and it easy to miss.

Additionally, I think that the same operations should also be performed in the select case above at line 968, otherwise the watcher is not cleaned up properly if the context is closed at that point.

oblazek replied:
> Additionally, I think that the same operations should also be performed in the select case above at line 968, otherwise the watcher is not cleaned up properly if the context is closed at that point.

yeah you are right.. 

> Do you think that it could be possible to preserve the handling of these two cases here, falling though to the stopWatch case?

yeah that would work as well.. (I am thinking `goto abort` in all these cases)

giorio94 replied:
I would suggest to just defer the cleanup logic, so that we are sure it is never missed (there is also another return case which is currently not handled).

oblazek replied:
right, that's even better, thanks

giorio94 reviewed (commented):
Thanks @oblazek!

The controller change LGTM. I left you a comment/question inline about the watcher termination handling.
